### PR TITLE
feat: add geminiAI

### LINKED ISSUE #23: feat: implement test in package chat

### LINKED ISSUE #36: feat: Gemini AI

### PR DESCRIPTION
**Pull request by nicholasboari**
## Getting Started

- Get an API key https://ai.google.dev/gemini-api/docs/api-key?hl=pt-br
- Go to application.yml and set your key
image
 

## PR Checklist

Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines
- [X] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

Issue Number: #36 

## PR Type

What kind of change does this PR introduce?

Add GeminiAI service

- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

### CODE REVIEW

**Review by rafael-dev2021 — CHANGES_REQUESTED** [review]
The code needs more testing.

**Review by ygorsimoes — CHANGES_REQUESTED** [review]
....

File: src/main/java/org/instituteatri/deep/service/ChatService.java [hunk 1]
```diff
@@ -1,52 +1,126 @@
...
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import lombok.RequiredArgsConstructor;
 import org.instituteatri.deep.dto.request.ChatRequestDTO;
+import org.instituteatri.deep.dto.response.ChatResponseDTO;
 import org.instituteatri.deep.dto.response.OccurrenceResponseDTO;
 import org.instituteatri.deep.model.Occurrence;
+import org.instituteatri.deep.model.Role;
 import org.modelmapper.ModelMapper;
 import org.slf4j.Logger;
```
Comment (ygorsimoes) [inline src/main/java/org/instituteatri/deep/service/ChatService.java:12]: Test lib

**Review by ygorsimoes — CHANGES_REQUESTED** [review]

File: src/main/java/org/instituteatri/deep/infrastructure/security/AppConfigTemplate.java [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package org.instituteatri.deep.infrastructure.security;
+
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.web.client.RestTemplate;
+
+@Configuration
+public class AppConfigTemplate {
```
Comment (rafael-dev2021) [inline src/main/java/org/instituteatri/deep/infrastructure/security/AppConfigTemplate.java]: Create config package within the infrastructure package for the configuration classes.

File: src/main/java/org/instituteatri/deep/service/ChatService.java [hunk 3]
```diff
@@ -1,52 +1,126 @@
...
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ChatService.class);
     private final OccurrenceService occurrenceService;
     private final MessageService messageService;
     private final ModelMapper modelMapper;
+    private final RestTemplate restTemplate;
 
-    public OllamaApi.ChatResponse generate(ChatRequestDTO request, String occurrenceId) {
+    public ChatResponse generate(ChatRequestDTO request, String occurrenceId) {
         OccurrenceResponseDTO occurrenceResponse = occurrenceService.getById(occurrenceId);
```
Comment (rafael-dev2021) [inline src/main/java/org/instituteatri/deep/service/ChatService.java:44]: Code repetition on lines 44-55 and 66-67

File: src/main/java/org/instituteatri/deep/service/ChatService.java [hunk 4]
```diff
@@ -1,52 +1,126 @@
...
+import org.springframework.web.client.RestTemplate;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
 
 @Service
 @RequiredArgsConstructor
 public class ChatService {
```
Comment (rafael-dev2021) [inline src/main/java/org/instituteatri/deep/service/ChatService.java:29]: Rename the methods of the class to make it clearer and more objective what they do.

File: src/main/java/org/instituteatri/deep/service/ChatService.java [hunk 5]
```diff
@@ -1,52 +1,126 @@
...
+    @Value("${gemini.key}")
+    String GOOGLE_API_KEY;
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ChatService.class);
     private final OccurrenceService occurrenceService;
     private final MessageService messageService;
     private final ModelMapper modelMapper;
+    private final RestTemplate restTemplate;
 
-    public OllamaApi.ChatResponse generate(ChatRequestDTO request, String occurrenceId) {
```
Comment (rafael-dev2021) [inline src/main/java/org/instituteatri/deep/service/ChatService.java:31]: Dividing responsibilities into smaller, clearer methods can improve the readability and maintainability of the code.

File: src/main/java/org/instituteatri/deep/service/ChatService.java [hunk 6]
```diff
@@ -1,52 +1,126 @@
...
         ChatResponse response = new OllamaApi().chat(build);
         messages.add(Message.builder(Message.Role.ASSISTANT)
                 .withContent(response.message().content()).build());
         LOGGER.info("Inserting {} to the database", msg);
-        messageService.save(response.message(), occurrenceId);
+        messageService.saveOllama(response.message(), occurrenceId);
         return response;
     }
+
+    public ChatResponseDTO generateGemini(ChatRequestDTO request, String occurrenceId) {
```
Comment (rafael-dev2021) [inline src/main/java/org/instituteatri/deep/service/ChatService.java]: Dividing responsibilities into smaller, clearer methods can improve the readability and maintainability of the code.

File: src/main/java/org/instituteatri/deep/service/ChatService.java [hunk 7]
```diff
@@ -1,52 +1,126 @@
...
+            // Parse to json and get text
+            text = getString(responseBody, text);
+            org.instituteatri.deep.model.Message msgAI = org.instituteatri.deep.model.Message.builder()
+                    .role(Role.MODEL)
+                    .occurrence(occurrence)
+                    .content(text)
+                    .build();
+            messageService.saveGemini(msgAI);
+        } else {
+            System.out.println("Request failed with status code: " + response.getStatusCode());
```
Comment (rafael-dev2021) [inline src/main/java/org/instituteatri/deep/service/ChatService.java]: Replace System.out.println with LOGGER.error or a custom exception to maintain consistency and improve error control.

File: src/main/java/org/instituteatri/deep/service/ChatService.java (hunk 4 above)
Comment (rafael-dev2021) [inline src/main/java/org/instituteatri/deep/service/ChatService.java:29]: Improve the identification of methods with line breaks.

File: src/main/java/org/instituteatri/deep/service/ChatService.java [hunk 8]
```diff
@@ -1,52 +1,126 @@
...
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
 
 @Service
 @RequiredArgsConstructor
 public class ChatService {
 
     @Value("${spring.ai.ollama.chat.options.model}")
     private String model;
```
Comment (rafael-dev2021) [inline src/main/java/org/instituteatri/deep/service/ChatService.java:32]: Only declared property?

File: src/main/java/org/instituteatri/deep/service/MessageService.java
Comment (rafael-dev2021) [inline src/main/java/org/instituteatri/deep/service/MessageService.java:1]: Just a note: if this MessageService class doesn't scale in the future of the project, wouldn't it be more feasible to create the saveOllama and saveGemini methods within the ChatService class itself?
Just an observation!

File: src/test/java/org/instituteatri/deep/controller/ChatControllerTest.java
Comment (rafael-dev2021) [inline src/test/java/org/instituteatri/deep/controller/ChatControllerTest.java:1]: If possible, add 2 tests for each endpoint, 1 success and 1 error.

File: src/test/java/org/instituteatri/deep/service/ChatServiceTest.java
Comment (rafael-dev2021) [inline src/test/java/org/instituteatri/deep/service/ChatServiceTest.java:1]: Change the sentences from Portuguese to English to meet the project's development standards and, if possible, add two tests for each method.

**Review by rafael-dev2021 — CHANGES_REQUESTED** [review]

File: src/main/java/org/instituteatri/deep/model/Role.java [hunk 9]
```diff
@@ -0,0 +1,6 @@
+package org.instituteatri.deep.model;
+
+public enum Role {
```
Comment (rafael-dev2021) [inline src/main/java/org/instituteatri/deep/model/Role.java]: Add a clear and objective name for the enum; just "Role" is too generic.

**Review by rafael-dev2021 — CHANGES_REQUESTED** [review]

**Review by rafael-dev2021 — APPROVED** [review]
LGTM

**Review by ffx64 — APPROVED** [review]
LGTM

**Review by george-prado — APPROVED** [review]
LGTM.

**Review by ygorsimoes — APPROVED** [review]
Good!

**Pull request merged**